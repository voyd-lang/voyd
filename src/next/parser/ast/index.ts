export * from "./expr.js";
export * from "./form.js";
export * from "./atom.js";
export * from "../reader-macros/lib/init-helpers.js";
export { is, SourceLocation } from "./syntax.js";
