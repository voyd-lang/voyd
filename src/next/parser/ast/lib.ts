import { FloatAtom, IdentifierAtom, IntAtom } from "./atom.js";
import { Expr } from "./expr.js";
import {
  ArrayLiteralForm,
  CallForm,
  FormElementInitVal,
  LabelForm,
  ObjectLiteralForm,
  ParenForm,
  TupleForm,
} from "./form.js";
import { is } from "./index.js";

export const call = (
  fn: IdentifierAtom | string,
  ...args: FormElementInitVal
) => new CallForm([fn, ...args]);

export const paren = (...args: FormElementInitVal) => new ParenForm(args);

export const tuple = (...args: FormElementInitVal) => new TupleForm(args);

export const arrayLiteral = (...args: FormElementInitVal) =>
  new ArrayLiteralForm(args);

export const objectLiteral = (...args: LabelForm[]) =>
  new ObjectLiteralForm(args);

export const label = (label: IdentifierAtom | string, value: Expr) =>
  new LabelForm([label, value]);

export const identifier = (id: string) => new IdentifierAtom(id);

export const int = (value: string | number, type: "i32" | "i64" = "i32") =>
  new IntAtom(String(value)).setType(type);

export const float = (value: string | number, type: "f32" | "f64" = "f64") =>
  new FloatAtom(String(value)).setType(type);

export const string = (value: string) => {
  const codes = value.split("").map((c) => int(c.charCodeAt(0)));

  return call(
    "new_string",
    call(
      "object",
      label("from", call("FixedArray", call("generics", "i32"), ...codes))
    )
  );
};

export const idIs = (id?: unknown, value: string) =>
  is(id, IdentifierAtom) && id.value === value;
