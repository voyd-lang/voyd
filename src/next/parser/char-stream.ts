import { SourceLocation } from "./ast/syntax.js";

export class CharStream {
  readonly filePath: string;
  readonly value: string;
  readonly length: number;
  readonly location = {
    index: 0,
    line: 1,
    column: 0,
  };

  constructor(contents: string, filePath: string) {
    this.value = contents;
    this.length = contents.length;
    this.filePath = filePath;
  }

  /** Current index the file is on */
  get position() {
    return this.location.index;
  }

  get line() {
    return this.location.line;
  }

  get column() {
    return this.location.column;
  }

  get hasCharacters() {
    return this.position < this.length;
  }

  get next() {
    return this.value[this.position];
  }

  currentSourceLocation() {
    const startLine = this.line;
    const startColumn = this.column;
    return new SourceLocation({
      startIndex: this.position,
      endIndex: this.position,
      startLine,
      endLine: startLine,
      startColumn,
      endColumn: startColumn,
      filePath: this.filePath,
    });
  }

  at(index: number): string | undefined {
    return this.value[this.position + index];
  }

  /** Returns the next character and advances the stream */
  consumeChar(): string {
    const char = this.next;
    if (char === undefined) {
      throw new Error("Out of characters");
    }

    this.location.index += 1;
    this.location.column += 1;
    if (char === "\n") {
      this.location.line += 1;
      this.location.column = 0;
    }

    return char;
  }
}
