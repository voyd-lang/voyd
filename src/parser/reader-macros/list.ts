import { ReaderMacro } from "./types.js";

export const listReader: ReaderMacro = {
  match: (t) => t.value === "(" || t.value === ")",
  macro: (file, { token, reader }) => {
    if (token.value === "(") {
      const list = reader(file, ")");
      list.setAttribute("tuple?", true);
      return list;
    }

    return undefined; // discard closing paren
  },
};
