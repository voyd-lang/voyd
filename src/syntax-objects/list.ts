import { Expr } from "./expr.js";
import { Float } from "./float.js";
import { getIdStr } from "./lib/get-id-str.js";
import { Id, Identifier } from "./identifier.js";
import { Int } from "./int.js";
import { SourceLocation, Syntax, SyntaxMetadata } from "./syntax.js";
import { ChildList } from "./lib/child-list.js";

type ListOpts =
  | ListValue[]
  | (SyntaxMetadata & {
      value?: ListValue[] | List | ChildList<Expr>;
      dynamicLocation?: boolean;
    });

export class List extends Syntax {
  readonly syntaxType = "list";
  #store = new ChildList(undefined, this);
  dynamicLocation?: boolean;

  constructor(opts: ListOpts) {
    opts = Array.isArray(opts) ? { value: opts } : opts;
    super(opts);
    const value = opts.value;
    this.dynamicLocation = opts.dynamicLocation;

    if (value instanceof Array) {
      value.forEach((v) => this.push(v));
    } else if (value instanceof List) {
      value.toArray().forEach((v) => this.push(v));
    }
  }

  get children() {
    return this.toArray();
  }

  get hasChildren() {
    return !!this.#store.length;
  }

  get length() {
    return this.#store.length;
  }

  at(index: number): Expr | undefined {
    return this.#store.at(index);
  }

  exprAt(index: number): Expr {
    const expr = this.#store.at(index);
    if (!expr) {
      throw new Error(`No expr at ${index}`);
    }
    return expr;
  }

  identifierAt(index: number): Identifier {
    const id = this.at(index);
    if (!id?.isIdentifier()) {
      throw new Error(`No identifier at index ${index}`);
    }
    return id;
  }

  optionalIdentifierAt(index: number): Identifier | undefined {
    const id = this.at(index);
    if (id?.isIdentifier()) return id;
  }

  listAt(index: number): List {
    const id = this.at(index);
    if (!id?.isList()) {
      throw new Error(`No list at index ${index}`);
    }
    return id;
  }

  set(index: number, expr: Expr | string) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    this.#store.set(index, result);
    return this;
  }

  calls(fnId: Id, atIndex = 0) {
    return this.getIdStrAt(atIndex) === getIdStr(fnId);
  }

  getIdStrAt(index: number): string | undefined {
    const v = this.at(index);
    return v?.isIdentifier() || v?.isStringLiteral() ? v.value : undefined;
  }

  consume(): Expr {
    return this.#store.consume();
  }

  first(): Expr | undefined {
    return this.#store.at(0);
  }

  last(): Expr | undefined {
    return this.#store.at(-1);
  }

  /** Returns all but the first element in an array */
  argsArray(): Expr[] {
    return this.#store.toArray().slice(1);
  }

  pop(): Expr | undefined {
    return this.#store.pop();
  }

  push(...expr: ListValue[]) {
    expr.forEach((ex) => {
      if (typeof ex === "string") {
        this.#store.push(new Identifier({ value: ex, parent: this }));
        return;
      }

      if (typeof ex === "number" && Number.isInteger(ex)) {
        this.#store.push(new Int({ value: ex, parent: this }));
        return;
      }

      if (typeof ex === "number") {
        this.#store.push(new Float({ value: ex, parent: this }));
        return;
      }

      if (ex instanceof Array) {
        this.push(new List({ value: ex, parent: this }));
        return;
      }

      if (ex.syntaxType === "nop") {
        return;
      }

      if (this.dynamicLocation) {
        this.updateLocationFrom(ex);
      }

      this.#store.push(ex);
    });

    return this;
  }

  private updateLocationFrom(ex: Expr) {
    if (!ex.location) return;

    if (!this.location) {
      this.location = ex.location.clone();
    }

    if (this.location.startIndex > ex.location.startIndex) {
      const nl = ex.location.clone();
      nl.setEndToEndOf(this.location);
      this.location = nl;
    }

    if (
      !this.location.endColumn ||
      this.location.endIndex < ex.location.endIndex
    ) {
      this.location?.setEndToEndOf(ex.location);
    }
  }

  findIndex(cb: (expr: Expr) => boolean) {
    return this.toArray().findIndex(cb);
  }

  insert(expr: Expr | string, at = 0) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    this.#store.insert(result, at);
    return this;
  }

  remove(index: number, count = 1) {
    this.#store.remove(index, count);
    return this;
  }

  each(fn: (expr: Expr, index: number, array: Expr[]) => void): List {
    this.toArray().forEach(fn);
    return this;
  }

  map(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.toArray().map(fn),
    });
  }

  flatMap(
    fn: (expr: Expr, index: number, array: Expr[]) => Expr | Expr[]
  ): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.toArray().flatMap(fn),
    });
  }

  reduce(
    fn: (acc: List, expr: Expr, index: number, array: Expr[]) => List
  ): List {
    return this.toArray().reduce(fn, new List({ ...super.getCloneOpts() }));
  }

  slice(start?: number, end?: number): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.#store.slice(start, end),
    });
  }

  sliceAsArray(start?: number, end?: number) {
    return this.children.slice(start, end);
  }

  /**
   * Returns the expression for a labeled argument, eg. for the list
   * `[hello, [:, world, 1]]` – which represents `hello(world: 1)` –
   * `labeledArg("world")` will return the expression `1`.
   *
   * If the label is not found, `undefined` is returned.
   */
  optionalLabeledArg(label: string): Expr | undefined {
    for (const expr of this.#store.toArray()) {
      if (!expr.isCall()) continue;
      if (!expr.calls(":")) continue;

      const labelId = expr.argAt(0);
      if (!labelId?.isIdentifier()) continue;

      if (labelId.value === label) {
        return expr.argAt(1);
      }
    }

    return undefined;
  }

  /** Asserts that a labeled argument exists and returns its expression. */
  labeledArg(label: string): Expr {
    const expr = this.optionalLabeledArg(label);
    if (!expr) throw new Error(`Labeled argument '${label}' not found`);
    return expr;
  }

  /**
   * Returns `true` when the list contains a labeled argument with the provided
   * name.
   */
  hasLabeledArg(label: string): boolean {
    return this.optionalLabeledArg(label) !== undefined;
  }

  /**
   * Builds and returns a map of all labeled arguments contained in the list.
   */
  getLabeledArgs(): Map<string, Expr> {
    const map = new Map<string, Expr>();
    for (const expr of this.#store.toArray()) {
      if (!expr.isCall() || !expr.calls(":")) continue;
      const labelId = expr.argAt(0);
      if (!labelId?.isIdentifier()) continue;
      const valueExpr = expr.argAt(1);
      if (valueExpr) map.set(labelId.value, valueExpr);
    }
    return map;
  }

  toArray(): Expr[] {
    return this.#store.toArray();
  }

  toJSON() {
    return this.toArray();
  }

  /** Clones should return a deep copy of all expressions except for type expressions */
  clone(parent?: Expr): List {
    return new List({
      ...super.getCloneOpts(parent),
      value: this.#store.clone(),
    });
  }
}

export type ListValue = Expr | string | number | ListValue[];
