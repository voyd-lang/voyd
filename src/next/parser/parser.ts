import { Expr } from "./ast/expr.js";
import { CharStream } from "./char-stream.js";
import { read } from "./reader.js";
import { expandSyntaxMacros } from "./syntax-macros/index.js";

export const parse = (text: string, filePath?: string): Expr => {
  const chars = new CharStream(text, filePath ?? "raw");
  const rawAst = read(chars);
  return expandSyntaxMacros(rawAst);
};
