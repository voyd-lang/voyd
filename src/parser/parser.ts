import { List } from "../syntax-objects/index.js";
import { CharStream } from "./char-stream.js";
import { parseChars } from "./parse-chars.js";
import { ReaderMacro } from "./reader-macros/types.js";
import { expandSyntaxMacros } from "./syntax-macros/index.js";

export const parse = (text: string, filePath?: string): List => {
  const chars = new CharStream(text, filePath ?? "raw");
  const rawAst = parseChars(chars);
  return expandSyntaxMacros(rawAst);
};
