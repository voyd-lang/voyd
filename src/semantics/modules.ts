import { ParsedModule, stdPath } from "../parser/index.js";
import { List } from "../syntax-objects/list.js";
import { RootModule, VoydModule } from "../syntax-objects/module.js";

/** Registers submodules of a parsed module for future import resolution */
export const registerModules = (
  opts: ParsedModule,
  rootModule: VoydModule = new RootModule({})
): VoydModule => {
  const { srcPath, files } = opts;

  for (const [filePath, file] of Object.entries(files)) {
    const resolvedPath = filePathToModulePath(
      filePath,
      srcPath ?? opts.indexPath
    );

    const parsedPath = resolvedPath.split("/").filter(Boolean);

    registerModule({
      path: parsedPath,
      parentModule: rootModule,
      ast: file.slice(1), // Skip the first element (ast)
      isIndex: filePath === opts.indexPath,
    });
  }

  return rootModule;
};

const registerModule = ({
  path,
  parentModule,
  ast,
  isIndex,
}: {
  path: string[];
  parentModule: VoydModule;
  ast: List;
  isIndex?: boolean;
}): VoydModule | undefined => {
  const [name, ...rest] = path;

  if (!name) return;

  const existingModule = parentModule.resolveEntity(name);

  if (existingModule && !existingModule.isModule()) {
    throw new Error(
      `Cannot register module ${name} because it is already registered as ${existingModule.syntaxType}`
    );
  }

  if (!existingModule && (name === "index" || name === "mod")) {
    parentModule.value = ast.toArray();
    registerDefaultImports(parentModule);
    parentModule.getAllEntities().forEach((e) => {
      if (e.isModule()) parentModule.push(e);
    });
    return;
  }

  const module =
    existingModule ??
    new VoydModule({
      ...(!rest.length ? { ...ast.metadata, value: ast.toArray() } : {}),
      name,
      isIndex,
    });

  if (!existingModule) {
    parentModule.push(module);
    registerDefaultImports(module);
  }

  if (!existingModule && (module.name.is("src") || module.name.is("std"))) {
    parentModule.registerExport(module);
  }

  if (existingModule && !rest.length) {
    module.unshift(...ast.toArray().reverse());
  }

  if (!rest.length) return;

  return registerModule({ path: rest, parentModule: module, ast });
};

const filePathToModulePath = (filePath: string, srcPath: string) => {
  let finalPath = filePath.startsWith(stdPath)
    ? filePath.replace(stdPath, "std")
    : filePath;

  finalPath = finalPath.startsWith(srcPath)
    ? finalPath.replace(srcPath, "src")
    : finalPath;

  finalPath = finalPath.replace(".voyd", "");

  return finalPath;
};

const registerDefaultImports = (module: VoydModule) => {
  module.unshift(new List(["use", ["::", "root", "all"]]));
  const mod = module.resolveModule("std");
  if (mod) return;
  module.unshift(new List(["use", ["::", "std", "all"]]));
};
