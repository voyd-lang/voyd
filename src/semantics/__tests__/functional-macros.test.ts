import { parse } from "../../parser/parser.js";
import path from "node:path";
import { registerModules } from "../modules.js";
import { expandFunctionalMacros } from "../functional-macros.js";
import { functionalMacrosVoydFile } from "./fixtures/functional-macros-voyd-file.js";
import { test } from "vitest";
import { List } from "../../syntax-objects/list.js";

test("functional macro expansion", async (t) => {
  const parserOutput = parse(functionalMacrosVoydFile);
  const files = {
    std: new List([]),
    test: parserOutput,
  };
  const resolvedModules = registerModules({
    files,
    srcPath: path.dirname("test"),
    indexPath: "test.voyd",
  });
  const result = expandFunctionalMacros(resolvedModules);
  t.expect(result).toMatchSnapshot();
});

test("quote2 expands macros within macros", async (t) => {
  const source = `
pub macro \`() quote2 $@body

pub macro binaryen_gc_call(func, args, return_type)
  \` binaryen func: $func namespace: gc args: $args return_type: $return_type

pub macro bin_type_to_heap_type(type)
  binaryen_gc_call(modBinaryenTypeToHeapType, \` [BnrType<($type)>])

bin_type_to_heap_type(FixedArray<Int>)
`;
  const parserOutput = parse(source);
  const files = { std: new List([]), test: parserOutput };
  const resolvedModules = registerModules({
    files,
    srcPath: path.dirname("test"),
    indexPath: "test.voyd",
  });
  const result = expandFunctionalMacros(resolvedModules);
  t.expect(result).toMatchSnapshot();
});
