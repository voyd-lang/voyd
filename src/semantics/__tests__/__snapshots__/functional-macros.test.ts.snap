// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functional macro expansion 1`] = `
[
  "module",
  "root",
  [
    "exports",
    [
      "std#902",
    ],
  ],
  [
    [
      "module",
      "std",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
      ],
    ],
    [
      "module",
      "test",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "functional-macro",
          "\`#931",
          [
            "parameters",
          ],
          [
            "block",
            [
              "block",
              [
                "quote",
                "quote",
                [
                  "$@",
                  "body",
                ],
              ],
            ],
          ],
        ],
        [
          "functional-macro",
          "let#978",
          [
            "parameters",
          ],
          [
            "block",
            [
              "block",
              [
                "define",
                "equals_expr",
                [
                  "extract",
                  "body",
                  0,
                ],
              ],
              [
                "quote",
                "define",
                [
                  "$",
                  [
                    "extract",
                    "equals_expr",
                    1,
                  ],
                ],
                [
                  "$",
                  [
                    "extract",
                    "equals_expr",
                    2,
                  ],
                ],
              ],
            ],
          ],
        ],
        [
          "define-macro-variable",
          "extract_parameters",
          [
            "reserved-for-type",
          ],
          [
            "is-mutable",
            false,
          ],
        ],
        [
          "functional-macro",
          "fn#1771",
          [
            "parameters",
          ],
          [
            "block",
            [
              "block",
              [
                "define",
                "first",
                [
                  "extract",
                  "body",
                  0,
                ],
              ],
              [
                "define",
                "is_equals",
                [
                  "==",
                  [
                    "extract",
                    "first",
                    0,
                  ],
                  "=",
                ],
              ],
              [
                "define",
                "definitions",
                [
                  "block",
                  [
                    "if",
                    "is_equals",
                    [
                      ":",
                      "then",
                      [
                        "block",
                        [
                          "extract",
                          "first",
                          1,
                        ],
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "block",
                        "first",
                      ],
                    ],
                  ],
                ],
              ],
              [
                "define",
                "identifier_list",
                [
                  "block",
                  [
                    "if",
                    [
                      "==",
                      [
                        "extract",
                        "definitions",
                        0,
                      ],
                      ":",
                    ],
                    [
                      ":",
                      "then",
                      [
                        "block",
                        [
                          "extract",
                          "definitions",
                          1,
                        ],
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "block",
                        [
                          "if",
                          [
                            "==",
                            [
                              "extract",
                              "definitions",
                              0,
                            ],
                            "->",
                          ],
                          [
                            ":",
                            "then",
                            [
                              "block",
                              [
                                "extract",
                                "definitions",
                                1,
                              ],
                            ],
                          ],
                          [
                            ":",
                            "else",
                            [
                              "block",
                              "definitions",
                            ],
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
              ],
              [
                "define",
                "return_type",
                [
                  "block",
                  [
                    "if",
                    [
                      "==",
                      [
                        "extract",
                        "definitions",
                        0,
                      ],
                      ":",
                    ],
                    [
                      ":",
                      "then",
                      [
                        "block",
                        [
                          "slice",
                          "definitions",
                          2,
                          3,
                        ],
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "block",
                        [
                          "if",
                          [
                            "==",
                            [
                              "extract",
                              "definitions",
                              0,
                            ],
                            "->",
                          ],
                          [
                            ":",
                            "then",
                            [
                              "block",
                              [
                                "slice",
                                "definitions",
                                2,
                                3,
                              ],
                            ],
                          ],
                          [
                            ":",
                            "else",
                            [
                              "quote",
                            ],
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
              ],
              [
                "define",
                "identifier",
                [
                  "extract",
                  "identifier_list",
                  0,
                ],
              ],
              [
                "define",
                "params",
                [
                  "extract_parameters",
                  "identifier_list",
                ],
              ],
              [
                "define",
                "expressions",
                [
                  "block",
                  [
                    "if",
                    "is_equals",
                    [
                      ":",
                      "then",
                      [
                        "block",
                        [
                          "slice",
                          "first",
                          2,
                        ],
                      ],
                    ],
                    [
                      ":",
                      "else",
                      [
                        "block",
                        [
                          "slice",
                          "body",
                          1,
                        ],
                      ],
                    ],
                  ],
                ],
              ],
              [
                "quote",
                "define_function",
                [
                  "$",
                  "identifier",
                ],
                [
                  "$",
                  "params",
                ],
                [
                  "return_type",
                  [
                    "$@",
                    "return_type",
                  ],
                ],
                [
                  "$",
                  [
                    "concat",
                    [
                      "quote",
                      "block",
                    ],
                    "expressions",
                  ],
                ],
              ],
            ],
          ],
        ],
        [
          "define_function",
          "fib",
          [
            "parameters",
            [
              ":",
              "n",
              "i32",
            ],
          ],
          [
            "return_type",
            "i32",
          ],
          [
            "block",
            [
              "block",
              [
                "define",
                "base",
                1,
              ],
              [
                "if",
                [
                  "<=",
                  "n",
                  "base",
                ],
                [
                  ":",
                  "then",
                  [
                    "block",
                    "n",
                  ],
                ],
                [
                  ":",
                  "else",
                  [
                    "block",
                    [
                      "+",
                      [
                        "fib",
                        [
                          "-",
                          "n",
                          1,
                        ],
                      ],
                      [
                        "fib",
                        [
                          "-",
                          "n",
                          2,
                        ],
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
  ],
]
`;

exports[`quote2 expands macros within macros 1`] = `
[
  "module",
  "root",
  [
    "exports",
    [
      "std#2384",
    ],
  ],
  [
    [
      "module",
      "std",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
      ],
    ],
    [
      "module",
      "test",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "pub",
          "macro",
          [
            "\`",
          ],
          "quote2",
          [
            "$@",
            "body",
          ],
        ],
        [
          "pub",
          "macro",
          [
            "binaryen_gc_call",
            "func",
            "args",
            "return_type",
          ],
          [
            "block",
            [
              "\`",
              "binaryen",
              [
                ":",
                "func",
                [
                  "$",
                  "func",
                ],
              ],
              [
                ":",
                "namespace",
                "gc",
              ],
              [
                ":",
                "args",
                [
                  "$",
                  "args",
                ],
              ],
              [
                ":",
                "return_type",
                [
                  "$",
                  "return_type",
                ],
              ],
            ],
          ],
        ],
        [
          "pub",
          "macro",
          [
            "bin_type_to_heap_type",
            "type",
          ],
          [
            "block",
            [
              "binaryen_gc_call",
              "modBinaryenTypeToHeapType",
              [
                "\`",
                [
                  "array",
                  [
                    "BnrType",
                    [
                      "generics",
                      [
                        "$",
                        "type",
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
        [
          "bin_type_to_heap_type",
          [
            "FixedArray",
            [
              "generics",
              "Int",
            ],
          ],
        ],
      ],
    ],
  ],
]
`;
